### PR TITLE
HPCC-20188 Fix 3 memory leaks

### PR DESCRIPTION
**Pull request by wangkx**
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@mckellyln please review.

File: system/jlib/jfile.cpp [hunk 1]
```diff
@@ -6239,6 +6239,13 @@ class CMemoryMappedFile: implements IMemoryMappedFile, public CInterface
         CloseHandle(hfile);
```
Comment (mckellyln) [inline system/jlib/jfile.cpp]: Just to be consistent, for WIN32, is UnmapViewOfFile() needed here or is it enough to CloseHandle(hmap) ?

**Comment by mckellyln** [comment]
@wangkx did you see these regions being released in valgrind output ?
Could you attach some before/after valgrind output here ?

File: system/jlib/jfile.cpp (hunk 1 above)
Comment (richardkchapman) [reply system/jlib/jfile.cpp]: @ghalliday Can you comment?

Comment (ghalliday) [reply system/jlib/jfile.cpp]: Good thinking, I think you are correct .  From the documentation:

"Mapped views of a file mapping object maintain internal references to the object, and a file mapping object does not close until all references to it are released. Therefore, to fully close a file mapping object, an application must unmap all mapped views of the file mapping object by calling UnmapViewOfFile and close the file mapping object handle by calling CloseHandle. These functions can be called in any order."

the same goes for the existing code in reinit() on line 6308.

@wangkx

**Review by mckellyln — APPROVED** [review]
I approve this change, provided we create a new JIRA for the Windows fix of calling UnmapViewOfFile()

**Comment by mckellyln** [comment]
https://track.hpccsystems.com/browse/HPCC-21069 created and linked to https://track.hpccsystems.com/browse/HPCC-20188

**Comment by mckellyln** [comment]
@ghalliday, @richardkchapman, @wangkx ok for merge

**Pull request merged**